exec/memory/base/src/main/java/org/apache/drill/exec/memory/DrillByteBufAllocator.java

```diff
@@ -111,31 +118,68 @@ public boolean isDirectBufferPooled() {
 
   @Override
   public ByteBuf heapBuffer() {
-    throw fail();
+    return HEAP_ALLOCATOR.allocateHeap();
   }
 
   @Override
   public ByteBuf heapBuffer(int initialCapacity) {
-    throw fail();
+    return HEAP_ALLOCATOR.allocateHeap(initialCapacity);
   }
 
   @Override
   public ByteBuf heapBuffer(int initialCapacity, int maxCapacity) {
-    throw fail();
+    return HEAP_ALLOCATOR.allocateHeap(initialCapacity, maxCapacity);
   }
 
   @Override
   public CompositeByteBuf compositeHeapBuffer() {
-    throw fail();
+    return compositeHeapBuffer(DEFAULT_MAX_COMPOSITE_COMPONENTS);
   }
 
   @Override
   public CompositeByteBuf compositeHeapBuffer(int maxNumComponents) {
-    throw fail();
-  }
-
-  private RuntimeException fail() {
-    throw new UnsupportedOperationException("Allocator doesn't support heap-based memory.");
+    return new CompositeByteBuf(this, false, maxNumComponents);
+  }
+
+  /**
+   * This method was copied from AbstractByteBufAllocator. Netty 4.1.x moved this method from
+   * AbstractByteBuf to AbstractByteBufAllocator. However, as DrillByteBufAllocator doesn't extend
+   * AbstractByteBufAllocator, it doesn't get the implementation automatically and we have to copy
+   * the codes.
+   */
+  @Override
+  public int calculateNewCapacity(int minNewCapacity, int maxCapacity) {
```

Comment:
Please remove this method copy and implementations above, and extend this class from `AbstractByteBufAllocator` (`newHeapBuffer` and `newDirectBuffer` implementations may be implemented as `HEAP_ALLOCATOR.allocateXXX`).